### DISCUSSION TITLE
How to serve a svelte-kit app via pocketbase public directory?

### DISCUSSION THREAD

**Discussion by cansati** [pocketbase/pocketbase#1098] [Q&A] [answered]
Hello everyone,

I'm trying to serve a prebuilded svelte-kit app by pocketbase. I've defined a public directory to serve a frontend (svelte) app by adding this code:

`e.Router.Get("/*", apis.StaticDirectoryHandler(os.DirFS("../pb_public"), false))`

 I have some questions.

1. Which svelte-kit adapter (vercel, netlify, cloudflare etc.) should i set to use via pocketbase?
2. How the file structure should be in the pb_public directory?
3. Is my approach to the goal OK? What would you suggest me?

**Comment by ganigeorgiev** [comment, marked as answer]
> Which svelte-kit adapter (vercel, netlify, cloudflare etc.) should i set to use via pocketbase?

The `pb_public` directory is suitable only for serving SPA or prerendered static content, so in your case I believe you should use the [static adapter](https://kit.svelte.dev/docs/adapters#supported-environments-static-sites), aka. `@sveltejs/adapter-static`.

---

> How the file structure should be in the pb_public directory?

This depends how you want to access it, but in the case of a static website you need every page to be generated as index.html in its own directory. This could be done by adding `trailingSlash: 'always'` to your `svelte.config.js`.

---

> Is my approach to the goal OK? What would you suggest me?

Yes, it is completely fine as long as you don't need server-side rendering (SSR).
As a simple example, you can also explore the code of the PocketBase site documentation in https://github.com/pocketbase/site (it is a static prerendered SvelteKit app).

**Comment by dsm-72** [comment]
@cansati that sounds super cool if it is what I think it is?
So you have a sveltekit app, and PocketBase as the DB. For deployment you then only use the static data from PocketBase and still get all the nice features like querying?
Can you share an example?

**Comment by dsm-72** [comment]
@cansati @ganigeorgiev can I get a bump on this?

**Reply by ganigeorgiev** [reply]
@dsm-72 I don't understand what is the request. Feel free to open a new discussion with more details what you are trying to do and I'll try to help.

**Comment by rodydavis** [comment]
Wanted to raise an alternative approach that works:

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
	proxy := httputil.NewSingleHostReverseProxy(&url.URL{
		Scheme: "http",
		Host:   "localhost:4321",
	})
	e.Router.Any("/*", echo.WrapHandler(proxy))
	e.Router.Any("/", echo.WrapHandler(proxy))
	return nil
})
```

This lets you proxy the request to a another url on the same server.

I go this working with Astro SSR and PocketBase.

**Reply by einarpersson** [reply]
Oooh. Out of curiosity, would this or something similar work if using js hooks also?  `httputil` doesn't show up in intellisense

**Reply by rodydavis** [reply]
It would work with Go hooks because the echo router library is needed!

https://rodydavis.com/posts/astro-ssr-pocketbase-single-server

**Reply by rodydavis** [reply]
Here is the full working code:

```go

package main

import (
	"log"
	"net/http/httputil"
	"net/url"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        proxy := httputil.NewSingleHostReverseProxy(&url.URL{
			Scheme: "http",
			Host:   "localhost:4321",
		})
		e.Router.Any("/*", echo.WrapHandler(proxy))
		e.Router.Any("/", echo.WrapHandler(proxy))
        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Reply by mlusetti** [reply]
I too have used in two cases this approach. BTW my feeling has always been that this is a "unwanted side effect". I mean this works and don't conflict with the /api and /_ prefix due to ordering, nothing else.
So I would like to ask @ganigeorgiev what he thought about this usage, there's the possibility that in a near future this will stop working or do you see this as a idiomatic usage for the framework ? 

Thanks a lot.

**Reply by ganigeorgiev** [reply]
@mlusetti About the routes - as long as there is no conflicting `/api/*` and `/_/*` routes it should work fine.

With that said, please note that full backward compatibility is not guaranteed before reaching stable v1.0.0 (_minor refactoring of the exposed internals is planned after the upcoming Presentator release in the next couple months_).

**Reply by mlusetti** [reply]
First of all thanks for the quick reply.
I know about 1.0.0 release and it is completely fine!
My main concern is that if we start to proxy /* to our frontend app (as per this code) and you refactor your internal code (totally legitimate) we could find ourselves in a bad situation.

**Reply by rodydavis** [reply]
This biggest risk would be if the internal router framework is swapped out (this can be resolved by using a different reverse proxy).

And this already uses a lower priority than the internal routes.

**Reply by rodydavis** [reply]
Also opened an issue about it:
https://github.com/pocketbase/pocketbase/issues/4276
